jena-arq/src/main/java/org/apache/jena/riot/system/AsyncParser.java

```diff
@@ -239,40 +278,143 @@ public void fatal(String message, long line, long col) {
                 throw new RiotException(SysRIOT.fmtMessage(message, line, col)) ;
             }
         };
+    }
+
+    /** Returns a runnable for stopping the parse process */
+    static Runnable startParserThread(
+            Logger LOG1, List<RDFParserBuilder> parserBuilders,
+            BlockingQueue<List<EltStreamRDFImpl>> queue,
+            int chunkSize,
+            boolean daemonMode) {
+
+        // A flag for whether parsing is considered aborted - regardless
+        // whether due to normal or exceptional conditions
+        AtomicBoolean isAborted = new AtomicBoolean(false);
+
+        // -- Parser thread setup
+        Consumer<List<EltStreamRDFImpl>> destination = batch -> {
+            if (!isAborted.get()) {
+                try {
+                    queue.put(batch);
+                } catch (InterruptedException ex) {
+                    // FmtLog.error(LOG, ex, "Error: %s", ex.getMessage());
+                    throw new RuntimeException(ex);
+                }
+            } else {
+                throw new RuntimeException(new InterruptedException());
+            }
+        };
+        EltStreamBatcher batcher = new EltStreamBatcher(destination, chunkSize);
+        StreamRDF generatorStream = new StreamToElements(batcher);
 
         // Parser thread
-        Runnable task = ()->{
-            batcher.startBatching();
+        Runnable task = () -> {
+            batcher.startBatching(); // No try/catch block currently needed because this is a no-op
```

Comment:
Personally, this seems odd. It is a no-op currently but implementation may change. What is the comment recording?

Reply:
If startBatching failed that it would have to be handled in a try/catch block. If it failed in the code as it is now then the resources wouldn't be freed; even the poison (end marker) would never be placed on the queue resulting in infinite wait.

Reply:
Revised in the factored out `Task` class.